### ISSUE TITLE
reproducible

### ISSUE THREAD

**Issue by Pluto1314** [ultralytics/yolov5#1222]
## ❔Question
Hello, my two training results on the custom datasets are different, but the random seed and other parameters have not been modified. What is going on?

## Additional context

1. First training
![图片](https://user-images.githubusercontent.com/47765770/97310010-b923df00-189d-11eb-8bd9-ceca57e35892.png)
2. Second training
![图片](https://user-images.githubusercontent.com/47765770/97310317-0a33d300-189e-11eb-9e54-b9a9e804c3e0.png)

I can be sure that my other settings and training commands have not changed.

**command:**
CUDA_VISIBLE_DEVICES=0,1,2,3 python -m torch.distributed.launch --nproc_per_node 4 train.py --batch-size 64 --data data/tiny_set.yaml --cfg yolov5s.yaml --weights '' --cache --device 0,1,2,3

**first training train_batch0.jpg**
![train_batch0](https://user-images.githubusercontent.com/47765770/97310837-ac53bb00-189e-11eb-8198-f71b5ec23ccc.jpg)

**second training train_batch0.jpg**
![train_batch0](https://user-images.githubusercontent.com/47765770/97310902-be355e00-189e-11eb-8bf4-969f9e2fe398.jpg)

**Comment by glenn-jocher** [comment]
@Pluto1314 there's not much use in examining partially completed trainings. 

In any case, it's common knowledge that multithreaded ops, and CUDA ops are typically non-deterministic in nature.
https://pytorch.org/docs/stable/notes/randomness.html

> There are some PyTorch functions that use CUDA functions that can be a source of nondeterminism. One class of such CUDA functions are atomic operations, in particular atomicAdd, which can lead to the order of additions being nondetermnistic. Because floating-point addition is not perfectly associative for floating-point operands, atomicAdd with floating-point operands can introduce different floating-point rounding errors on each evaluation, which introduces a source of nondeterministic variance (aka noise) in the result.
> 
> PyTorch functions that use atomicAdd in the forward kernels include torch.Tensor.index_add_(), torch.Tensor.scatter_add_(), torch.bincount().
> 
> A number of operations have backwards kernels that use atomicAdd, including torch.nn.functional.embedding_bag(), torch.nn.functional.ctc_loss(), torch.nn.functional.interpolate(), and many forms of pooling, padding, and sampling.
> 
> There is currently no simple way of avoiding nondeterminism in these functions.
> 
> Additionally, the backward path for repeat_interleave() operates nondeterministically on the CUDA backend because repeat_interleave() is implemented using index_select(), the backward path for which is implemented using index_add_(), which is known to operate nondeterministically (in the forward direction) on the CUDA backend (see above).

**Comment by glenn-jocher** [comment]
Also from https://pytorch.org/docs/stable/notes/randomness.html:
> REPRODUCIBILITY
> 
> Completely reproducible results are not guaranteed across PyTorch releases, individual commits or different platforms. Furthermore, results need not be reproducible between CPU and GPU executions, even when using identical seeds.

**Comment by Pluto1314** [comment]
Thank you very much for your response, so does this uncertainty have a big impact on the results?

**Comment by glenn-jocher** [comment]
@Pluto1314 variations due to inherent reproducibility issues would have the same impact on the final result as seed changes I suppose. In COCO training with different seeds would yield similar mAPs (but completely different model weights).

**Comment by glenn-jocher** [comment]
The only method I'm aware of that might guarantee you identical results might be to train on CPU with --workers 0, but this is impractical naturally, so you simply need to adapt your workflow to accommodate minor variations in final model results.

**Comment by Pluto1314** [comment]
I get it, thank you very much for your answers.

**Comment by Pluto1314** [comment]
I will close this issue, thanks again.

**Comment by Pluto1314** [comment]
Hi, sorry to disturb you again. My two testing results got a big difference.

1. First training weight
![图片](https://user-images.githubusercontent.com/47765770/97379995-143aed80-1901-11eb-8fca-0189d293ca3f.png)

2. Second training weight
![图片](https://user-images.githubusercontent.com/47765770/97380052-316fbc00-1901-11eb-9f37-adb549a3eec2.png)

**Comment by Pluto1314** [comment]
@glenn-jocher Will it not make such a big difference if I use a single GPU?

**Comment by glenn-jocher** [comment]
@Pluto1314 I have nothing to add to what I already posted before.

**Comment by dongjuns** [comment]
@Pluto1314 Hey, just experiment with it. Even you have awesome P100 GPUs, haha. :)

**Comment by Pluto1314** [comment]
@dongjuns I really don't know how to solve it, even if I add these:

![图片](https://user-images.githubusercontent.com/47765770/97768886-5a8e8780-1b61-11eb-90be-6e3e27ded03f.png)
![图片](https://user-images.githubusercontent.com/47765770/97768889-60846880-1b61-11eb-8418-90dd5cc4563f.png)

**Comment by hugh-98** [comment]
@Pluto1314 Hello! I have the same trouble recently. Do you know how to solve it?

**Comment by glenn-jocher** [comment]
@hugh-98 👋 Hello! Thanks for asking about **training reproducibility**. YOLOv5 🚀 uses a single training seed which is set here using the `init_seeds()` function.
https://github.com/ultralytics/yolov5/blob/6865d19a92d8c160c7fc3c92256627dadce1cd1e/train.py#L108

This function sets `python`, `numpy` and `torch` seeds and updates `cudnn` settings:
https://github.com/ultralytics/yolov5/blob/6865d19a92d8c160c7fc3c92256627dadce1cd1e/utils/general.py#L122-L130

To set a new seed for example:
```python
    init_seeds(2 + RANK)  # set seed=2
```

Note that even when using the same seed trainings may produce different results, especially when using CUDA backends. See https://pytorch.org/docs/stable/notes/randomness.html for details on factors affecting PyTorch training reproducibility and sources of randomness.

<img width="949" alt="Screen Shot 2022-01-08 at 11 20 59 AM" src="https://user-images.githubusercontent.com/26833433/148660366-9155dfd9-351c-4174-818f-2b1b907b3e4e.png">

Good luck 🍀 and let us know if you have any other questions!
